<launch> 
    
    <!-- Arguments, use this to set config parameters -->
    <arg name="topic_publish" value="/lkf2/pose" />
    <arg name="topic_pose_imu" value="/posePub_merged" />
    <arg name="topic_pose_cam" value="/camera/pose" />
    <arg name="frame_id_imu" value="imu_frame" />
    <arg name="frame_id_cam" value="camera_frame" />
    <arg name="imu_rate" value="125" /> <!-- in Hz -->
    <arg name="cam_rate" value="200" /> <!-- in Hz -->
    <arg name="sphere_radius" value="0.155" /> <!-- in m -->
    <arg name="debug_topics" value="true" /> 
    <!-- arg name="assume_flat_floor" value="true" -->
    <!-- arg name="max_upscaling_factor" value="true" -->

    <!-- Convert arguments to rosparam server -->
    <param name="topic_publish" type="string" value="$(arg topic_publish)" />
    <param name="topic_pose_imu" type="string" value="$(arg topic_pose_imu)" />
    <param name="topic_pose_cam" type="string" value="$(arg topic_pose_cam)" />
    <param name="frame_id_imu" type="string" value="$(arg frame_id_imu)" />
    <param name="frame_id_cam" type="string" value="$(arg frame_id_cam)" />
    <param name="imu_rate" type="int" value="$(arg imu_rate)" />
    <param name="cam_rate" type="int" value="$(arg cam_rate)" />
    <param name="sphere_radius" type="double" value="$(arg sphere_radius)" /> 
    <param name="debug_topics" type="bool" value="$(arg debug_topics)"/> 

    <!-- Arguments for the vizualiser node: pose to lidar -->
    <arg name="topic_listen" value="/lkf2/pose" />
    <arg name="global_frame" value="map3" />
    <arg name="pose_frame" value="odom" />
    <arg name="lidar_frame" value="pandar_frame" />

    <!-- Order of transforms: x y z rz ry rx, intrinsic rotation order: rx first, new ry second, new rz third -->

    <!-- Defines the transform between CAM and IMU coordinate system axes, e.g.: IMU x axis == CAM z axis -->
    <node pkg="tf" type="static_transform_publisher" name="left_cam_2_right_imu_stfb" args="0 0 0 1.570796 0 3.141592 axes_cam axes_imu 200" />
    <!-- Defines the transform between CAM and IMU coordinate systems in the global map frame, e.g.: IMU moves in x, CAM in y -->
    <node pkg="tf" type="static_transform_publisher" name="extr_imu_2_cam_stfb" args="0 0 0 -1.570796 0 0 $(arg frame_id_imu) $(arg frame_id_cam) 200" />
    <node pkg="tf" type="static_transform_publisher" name="lidar_2_odom_stfb" args="0 0 0 3.141592 0 0 $(arg pose_frame) $(arg lidar_frame) 200" />
    <node pkg="tf" type="static_transform_publisher" name="odom_is_imu_frame_stfb" args="0 0 0 0 0 0 $(arg pose_frame) $(arg frame_id_imu) 200" />

    <!-- node pkg="delta_pose_filter" type="filter_node" name="delta_pose_filter_node" output="screen" / -->

    <node pkg="delta_pose_filter" type="kalman_filter_node" name="lkf_pose_filter_node" output="screen" />

    <node pkg="delta_pose_filter" type="vizualizer" name="delta_tf_lidar_pub" output="screen" />

</launch>